plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -2241,7 +2241,8 @@ public ConnectorTableProperties getTableProperties(ConnectorSession session, Con
         if (handle.getPartitions().equals(newHandle.getPartitions()) &&
                 handle.getCompactEffectivePredicate().equals(newHandle.getCompactEffectivePredicate()) &&
                 handle.getBucketFilter().equals(newHandle.getBucketFilter()) &&
-                handle.getConstraintColumns().equals(newHandle.getConstraintColumns())) {
+                handle.getConstraintColumns().equals(newHandle.getConstraintColumns()) &&
+                partitionResult.getUnenforcedConstraint().equals(constraint.getSummary())) {
```

Comment:
I'm not sure this will stop. Consider predicate on `part` and `non_part` columns. Let's assume that
```
DomainTranslator.fromPredicate
```
returned
```
domain=[
  part = [0, 20],
  non_part = [0, 10]
],
remainingExpression = some_complex_filter_on_part
```
then in `applyFilter`
```
partitionResult.getUnenforcedConstraint()
```
would be:
```
unenforcedConstraint = [
  non_part = [0, 10]
]
```
since `remainingExpression` is not `true` and 
`!partitionResult.getUnenforcedConstraint().equals(constraint.getSummary())`
this rule would loop forever

Reply:
`constraint.getSummary()` won't contain the `cos(part)`

Reply:
> constraint.getSummary() won't contain the cos(part)

no, but it can contain best known summary for `cos(part)` which is `[0, inf]`

Reply:
Added a test for two test cases,

- with `substring` on  a non-part column
- with `% 2 = 0` on a part column

Reply:
This seem to work because of how `PushPredicateIntoTableScan` is implemented. `PushPredicateIntoTableScan#arePlansSame` returns `true` if table handles did not change.
However, if `PushPredicateIntoTableScan` relied on `applyFilter` to return `Optional.empty()` if nothing changed the rule would loop forever for `part LIKE 't%'` predicate.

This means that `HiveMetadata#applyFilter` does not meet SPI contract of returning `Optional.empty()` when nothing changed. This also means that we could remove entire:
```
        if (handle.getPartitions().equals(newHandle.getPartitions()) &&
                handle.getCompactEffectivePredicate().equals(newHandle.getCompactEffectivePredicate()) &&
                handle.getBucketFilter().equals(newHandle.getBucketFilter()) &&
                handle.getConstraintColumns().equals(newHandle.getConstraintColumns())) {
           return Optional.empty()
```
and the rule would not loop (because `TableHandles` stay the same).

IMO, we should write `applyFilter` in such a way that it meets SPI contract and not rely on actual `PushPredicateIntoTableScan` implementation.

Ideally, the engine is able to recognize that predicate above TS is redundant to TS `enforcedPredicate` (or `ConnectorTableProperties#predicate`) and not even try to push it down.

wdyt @martint

Reply:
i do not agree with the assessment that the implementation here violates the SPI contract. 
The connector is able to accept a predicate and subsume it, so it should return a non-empty result.

The fact that engine can _repeatedly_ ask a connector to accept same filter (despite the filter being subsumed each time), is _engine's_ responsibility.
And the truth is -- the engine diligently fulfills it via `PushPredicateIntoTableScan#arePlansSame`.

Reply:
Why to keep
```
        if (handle.getPartitions().equals(newHandle.getPartitions()) &&
                handle.getCompactEffectivePredicate().equals(newHandle.getCompactEffectivePredicate()) &&
                handle.getBucketFilter().equals(newHandle.getBucketFilter()) &&
                handle.getConstraintColumns().equals(newHandle.getConstraintColumns())) {
           return Optional.empty()
```
then if it's not effective and could be removed (the last predicate is not trivial)?

> The connector is able to accept a predicate and subsume it, so it should return a non-empty result.

This is the assumption of connector that there is a predicate that is being consumed. However, `Constraint` contains `summary` which doesn't imply that some predicate will be consumed when connector enforces it.

I guess it's more high level design question: Is it the engine responsibility to compare `ConnectorTableHandles` before/after pushdown or it's connector responsibility to return `empty` if nothing changed.

Reply:
The intent of the SPI contract is that if the application of a filter/projection/etc has "no effect", it should return `Optional.empty()`. From the javadocs:

> Connectors can indicate whether they don't support predicate pushdown or that the action had no effect
> by returning {@link Optional#empty()}. Connectors should expect this method to be called multiple times
> during the optimization of a given query.
> 
> <b>Note</b>: it's critical for connectors to return Optional.empty() if calling this method has no effect for that
> invocation, even if the connector generally supports pushdown. Doing otherwise can cause the optimizer
> to loop indefinitely.

The plansAreSame check is there due to legacy reasons, but we should eventually remove it. Comparing plans after every action can be very expensive, and the connector is in the best position to determine whether anything changed based on the semantics of the operation it's dealing with and whatever internal state is tracking. This is going to increasingly important as pushdown becomes more sophisticated (to the point of being able to push down very deep and complex query plans).

Reply:
> The intent of the SPI contract is that if the application of a filter/projection/etc has "no effect", it should return `Optional.empty()`. From the javadocs:
> [...] 

I read the javadoc and my interpretation of "no effect" was different than yours (maybe because i was biased towards a particular solution).
In particular, returning a remaining filter different  than  the input can be considered "an effect" (unless i am alone on this).

If we should close this PR because "no effect" definition is different, can you please update the javadoc to reflect that?

> The plansAreSame check is there due to legacy reasons, but we should eventually remove it.

The code as i read it today does not convey this intent while it easily could, perhaps with a code comment.